corpus_test.go

```diff
@@ -0,0 +1,187 @@
+		Resource  jsonEntity `json:"resource"`
+		Context   Record     `json:"context"`
+		Decision  string     `json:"decision"`
+		Reasons   []string   `json:"reason"`
+		Errors    []string   `json:"errors"`
+	} `json:"requests"`
+}
+
+type FileDataPointer struct {
+	Position int64
+	Size     int64
+}
+
+type FileDataMap struct {
+	buf   io.ReaderAt
+	files map[string]FileDataPointer
+}
+
+func NewFileDataMap(buf io.ReaderAt) FileDataMap {
+	return FileDataMap{
+		buf:   buf,
+		files: make(map[string]FileDataPointer),
+	}
+}
+
+func (fdm *FileDataMap) AddFileData(path string, position int64, size int64) {
+	fdm.files[path] = FileDataPointer{position, size}
+}
+
+func (fdm FileDataMap) GetFileData(path string) ([]byte, error) {
+	fdp := fdm.files[path]
+	content := make([]byte, fdp.Size)
+	_, err := fdm.buf.ReadAt(content, fdp.Position)
+	if err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+func TestCorpus(t *testing.T) {
+	t.Parallel()
+
+	corpusArchive, err := os.Open("corpus-tests.tar.gz")
+	if err != nil {
+		t.Fatal("error opening corpus archive", err)
+	}
+	defer corpusArchive.Close()
+
+	gzipReader, err := gzip.NewReader(corpusArchive)
+	if err != nil {
+		t.Fatal("error reading corpus compressed archive header", err)
+	}
+	defer gzipReader.Close()
+
+	buf, err := io.ReadAll(gzipReader)
+	if err != nil {
+		t.Fatal("error reading corpus compressed archive", err)
+	}
+
+	bufReader := bytes.NewReader(buf)
+	archiveReader := tar.NewReader(bufReader)
+
+	fdm := NewFileDataMap(bufReader)
+	var testFiles []string
+	for file, err := archiveReader.Next(); err == nil; file, err = archiveReader.Next() {
+		if file.Typeflag != tar.TypeReg {
+			continue
+		}
+
+		cursor, _ := bufReader.Seek(0, io.SeekCurrent)
+		fdm.AddFileData(file.Name, cursor, file.Size)
+
+		if strings.HasSuffix(file.Name, ".json") && !strings.HasSuffix(file.Name, ".entities.json") {
+			testFiles = append(testFiles, file.Name)
+		}
+	}
+
+	for _, testFile := range testFiles {
+		testFile := testFile
+		t.Run(testFile, func(t *testing.T) {
+			t.Parallel()
+			testContent, err := fdm.GetFileData(testFile)
+			if err != nil {
+				t.Fatal("error reading test content", err)
+			}
+
+			var tt corpusTest
+			if err := json.Unmarshal(testContent, &tt); err != nil {
+				t.Fatal("error unmarshalling test", err)
+			}
+
+			entitiesContent, err := fdm.GetFileData(tt.Entities)
+			if err != nil {
+				t.Fatal("error reading entities content", err)
+			}
+
+			var entities Entities
+			if err := json.Unmarshal(entitiesContent, &entities); err != nil {
+				t.Fatal("error unmarshalling test", err)
+			}
+
+			policyContent, err := fdm.GetFileData(tt.Policies)
+			if err != nil {
+				t.Fatal("error reading policy content", err)
+			}
+
+			policySet, err := NewPolicySet("policy.cedar", policyContent)
+			if err != nil {
+				t.Fatal("error parsing policy set", err)
+			}
+
+			for _, request := range tt.Requests {
+				t.Run(request.Desc, func(t *testing.T) {
```

Comment:
I'm not sure if the Go compiler will magically capture request or not in this case. Maybe check?
`request := request` was previously used when firing off goroutines.  This isn't required anymore, but I don't know if lambdas capture it or not.

Reply:
Yes, [this is mentioned](https://arc.net/l/quote/jlqattmy) in the blog post introducing the change to loop variables.

We can confirm that it's true in our case by printing out test spew with `go test . -v` and observing that the tests have unique names. For example:

```
    --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_4 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_6 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_2 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_3 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_1 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_0 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_7 (0.00s)
        --- PASS: TestCorpus/corpus-tests/395fa1e5e2d19129498bebcec8978e05d7b835b8.json/Request_5 (0.00s)
    --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_7 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_6 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_5 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_2 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_3 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_4 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_0 (0.00s)
        --- PASS: TestCorpus/corpus-tests/0a0c95def68ce2d6d91fe8935e96b38c98ca8940.json/Request_1 (0.00s)
```